### DISCUSSION TITLE
Customize WYSIWYG content - Options Override

### DISCUSSION THREAD

**Discussion by ZarckDev** [directus/directus#9960] [Archive]
Hi all,

Discovering and loving Directus so far!! Congrats for the work done with v9!!
Now, I get from discussions that Directus uses TinyMCE for WYSIWYG editor, and that we need to go through the `content_style` option to customize content css.

Here is an example of what I'm looking for:
I would like to apply margin bottom style to all `<p>` by default. In my case my front-end app uses TailwindCSS, so I just want to apply "`mb-2`" class to all paragraphs written in editor.

I just don't get **how** and **where** should I do that. I see that the interface in a WYSIWYG field allow to add 'Custom Formats' and 'Options Override' but do not know how to use that. 
<img src="https://user-images.githubusercontent.com/55784207/142771590-0517027b-47dc-4f01-a1a8-38ece04e01dd.png" width="400" height="400">

Is there a documentation about this that I missed?

Thank you for your help on this.

**Comment by rijkvanzanten** [comment]
Heya!

The "Custom Formats" option is what you're looking for. If you click the button in the top right of that interface, it'll pre-fill an example. That should help you get started. The documentation for what you can configure in there and what the exact format is can be found on https://www.tiny.cloud/docs/configure/content-formatting/#formats / https://www.tiny.cloud/docs/demo/format-custom/

**Comment by ZarckDev** [comment]
Thank you for you answer and documentation links.

So if we look at the pre-fill example from the "Custom Formats" area, it looks like we are filling `style_formats` array for `tinymce.init()` function, is it right? 

In my case (set a tailwind margin bottom class to all `<p>`), I set this into "Custom Formats":
```
[
    {
        "title": "Custom Paragraph",
        "selector": "p",
        "classes": "mb-2"
    }
]
```
And it works as long as we select "Custom paragraph" in the text area when creating content, because "Paragraph" is selected by default.

Here is my question then, when creating content, is there a way to take this custom format  by default? Or replace the default "Paragraph" format?
I was thinking about removing "title" from the Custom Formats to make it based on the "selector", but looks like it's needed (blank text area in content edition) :/


Thanks!

**Reply by rijkvanzanten** [reply]
Ah, you want to add mb-2 to _every_ paragraph by default? Normally, I'd say just go with CSS:

```css
#my-custom-content p {
  margin-bottom: 20px;
}
```

however, by using Tailwind this is not really what you'd use. I don't think there's a way to tell TinyMCE to add a class to every paragraph by default 🤔

**Reply by ZarckDev** [reply]
Yes as I'm using Tailwind a lot to avoid any custom CSS I try to make it as clean as possible :satisfied:.

According to tinyMCE documentation, we can customize default format [https://www.tiny.cloud/docs/configure/content-formatting/](https://www.tiny.cloud/docs/configure/content-formatting/) :

```
tinymce.init({
  selector: 'textarea',
  formats: {
    // Changes the default format for h1 to have a class of heading
    h1: { block: 'h1', classes: 'heading' }
  }
});
```

But what is exactly taking the "Custom Format" editor for the `tinyMCE.init()` function? is it the whole object or just `style_formats` key/value ?

-----------------

By the way, I realized this: 

1. we write some content with a custom format previously defined (the one above for example).
2. we modify the custom format , let's say instead of `mb-2`, we set `mb-5` or any class name.
3. back to the content, it does not automatically modify the class in the content previously set as "Custom Paragraph"
4. If we click it back to be "Custom Paragraph" again, it will add the class to the class list
See the image below
<img src="https://user-images.githubusercontent.com/55784207/142903723-7596ea13-280c-43a0-9459-78e873f48f75.png" width="450" height="200">

I don't know if it's an expected behaviour or not. I guess updating a custom format would overwrite an existing content's format.


Thanks

**Reply by rijkvanzanten** [reply]
> According to tinyMCE documentation, we can customize default format

Ah! In that case, you might have success by adding the following to "Options Override":

```json
{
  "formats": {
    "p": { "block": "p", "classes": "mb-2" }
  }
}
```

> But what is exactly taking the "Custom Format" editor for the tinyMCE.init() function? is it the whole object or just style_formats key/value ?

The "Custom Formats" option value is passed as-is to TinyMCE's `style_formats` object, so the structure is the same. 

> I don't know if it's an expected behaviour or not. I guess updating a custom format would overwrite an existing content's format.

Good question.. I'm not sure, this is all inner workings of TinyMCE, so I'm not too sure how it's supposed to handle this case.

**Reply by ZarckDev** [reply]
Thanks for the suggestion, but does not seem to work when adding this to "Options Override" (on Directus 9.0.1)

Ok, thanks for the precisions.

**Comment by lambdajack** [comment]
Hi, 

If it's just tailwindcss that you want this to work for you could try the following:

1. set the parent ID of WYSIWYG content

> `<div id="wysiwyg-card"></div>`

2. Set the children `<p>` of the above elements in your in your global css file (where you import tailwind:

```
@tailwind base;
@layer base {
  #wysiwyg-card p {
        @apply mb-2;
  }
}
@tailwind components;
@tailwind utilities;
```

I too am using tailwind the wysiwyg together and this solution reliably works for me.

**Reply by ZarckDev** [reply]
Ok, great! I will do this way then, thank you!
